### PR TITLE
feat: add edge middleware support to `ntl dev`

### PR DESCRIPTION
**Pull request by ascorbic**
### Summary
Adds support for running middleware using edge functions with `ntl dev`. 

When `next dev` is run, it doesn't emit any of the compiled files that we need to run middleware via our normal edge function. This means that currently we need to run `ntl build` before `ntl dev`, and changes would require a recompiled. This is not good DX: we want to be able to run `ntl dev` without building or linking the site, and have it update middleware when it is changed.

In this PR this is implemented with a special dev-only edge function which runs the middleware directly, rather than via the Next server. While importing the middleware.ts file directly works in simple cases, it breaks if the middleware imports anything else, because the imports won't be Deno-compatible. We handle this by running `esbuild --watch` on the middleware, and importing the compiled middleware. It still works if the middleware hasn't been created yet because esbuild will compile it when it is.

The wrapper edge function checks at request time whether the compiled middleware exists. If not, it just silently returns as there is no middleware. If it does exist then it dynamically imports it, then executes it. This means that it will always have the latest version without needing to reload the edge function.

The edge function is generated in a new `onDev` hook, which also spawns `esbuild`. A regular build will delete the dev edge function to ensure it's never deployed, and the dev hook deletes the production edge functions to ensure they don't conflict.

This needs e2e tests, which will follow later: #1553

### Test plan

1. Check out repo
2. Run npm i in root
3. Install latest version of CLI once [11.0.0](https://github.com/netlify/cli/pull/4970) is released
4. Edit `demos/middleware/netlify.toml` to switch plugin package names
5. Run `npm i` and `ntl dev` in middleware demo
6. Load site and check the "rewrite static" demo. You should see a cat and your city.

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal
capy lifeguard

### Standard checks:

- [ ] Check the Deploy Preview's Demo site for your PR's functionality
- [ ] Add docs when necessary

---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.

### CODE REVIEW

File: demos/middleware/middleware.ts [hunk 1]
```diff
@@ -58,6 +55,16 @@ export async function middleware(req: NextRequest) {
       response.headers.set('x-modified-in-rewrite', 'true')
     }
 
+    if (pathname.startsWith('/shows/redirectme')) {
```
Comment (ascorbic) [inline demos/middleware/middleware.ts:58]: Redirects weren't working, so this is a test for the fix

Comment (ascorbic) [inline demos/middleware/middleware.ts:58]: Redirects were broken, so these are the added tests

File: demos/middleware/netlify.toml [hunk 2]
```diff
@@ -4,7 +4,7 @@ publish = ".next"
 ignore = "if [ $CACHED_COMMIT_REF == $COMMIT_REF ]; then (exit 1); else git diff --quiet $CACHED_COMMIT_REF $COMMIT_REF ../..; fi;"
 
 [[plugins]]
-package = "../plugin-wrapper/"
+package = "@netlify/plugin-nextjs"
```
Comment (ascorbic) [inline demos/middleware/netlify.toml:7]: `ntl dev` won't de-dupe if the wrapper is used

File: packages/runtime/src/index.ts [hunk 3]
```diff
@@ -218,5 +219,29 @@ const plugin: NetlifyPlugin = {
     warnForRootRedirects({ appDir })
   },
 }
-module.exports = plugin
+// The types haven't been updated yet
+const nextRuntime = (
+  _inputs,
+  meta: { events?: Set<string> } = {},
+): NetlifyPlugin & { onPreDev?: NetlifyPlugin['onPreBuild'] } => {
+  if (!meta?.events?.has('onPreDev')) {
```
Comment (eduardoboucas) [inline packages/runtime/src/index.ts:227]: Great touch!

File: packages/runtime/src/helpers/dev.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import { resolve } from 'path'
+
+import { OnPreBuild } from '@netlify/build'
+import execa from 'execa'
+import { unlink, existsSync } from 'fs-extra'
+
+import { writeDevEdgeFunction } from './edge'
+import { patchNextFiles } from './files'
+
+// The types haven't been updated yet
```
Comment (eduardoboucas) [inline packages/runtime/src/helpers/dev.ts:10]: Ah, good point.

**Review by eduardoboucas — APPROVED** [review]
Didn't go deep into the Next.js parts, but LGTM!

**Pull request merged**